docs: add upgrade documentation (#1784)
defenseunicorns/uds-core: docs/reference/deployment/upgrades.md

> ---

Reviewer: This is great! Trying to think of how we can make this more obvious on the doc site, adding a link on the [landing page](https://uds.defenseunicorns.com/) `Working with UDS Core` would be a good start. Wondering if there's also value in a link on the [UDS Core Overview page](https://uds.defenseunicorns.com/reference/uds-core/overview/).

Author: I definitely like the landing page idea - I can make a follow-on docs PR to get it linked there. The overview (with its current content) doesn't feel like it has a good spot for this, but I do feel like it should be cross-linked in other places 🤔

Author: I have a draft PR ready for when this merges - https://github.com/defenseunicorns/uds-docs/pull/275.

title: Upgrading UDS Core
tableOfContents:
  maxHeadingLevel: 3
---

> This guide provides instructions and best practices for upgrading UDS Core installations. Following these guidelines helps ensure a smooth upgrade process and minimizes potential disruptions to your environment.

Reviewer 2: nit:
```suggestion
This guide provides instructions and best practices for upgrading UDS Core installations. Following these guidelines will help ensure a smooth upgrade process and minimize potential disruptions to your environment.
```

## Importance of Upgrades

Regularly upgrading UDS Core is essential for several reasons:

- **Security Patches**: Critical CVE fixes for both UDS Core components and underlying open source tooling
- **Bug Fixes**: Resolving issues in UDS Core and integrated open source components
- **New Features**: Access to new capabilities and improvements
- **Compatibility**: Ensuring continued compatibility with the broader UDS ecosystem

> Staying current with UDS Core releases helps maintain the security posture and functionality of your deployment.

Reviewer 2: nit: maybe environment is a better word to capture the state of the system?
```suggestion
Staying current with UDS Core releases helps maintain the security posture and functionality of your environment.
```

## Upgrade Strategies

### Sequential Minor Version Upgrades

UDS Core is designed and tested for sequential minor version upgrades (e.g., 0.9.0 → 0.10.0 → 0.11.0). This approach:

- Follows the tested upgrade path
- Allows for incremental validation of each upgrade step

> - Reduces complexity when troubleshooting issues

Reviewer 2: nit: slight tightening of language
```suggestion
- Reduces complexity during troubleshooting
```

- Perform the upgrade in staging first
- Validate all functionality before proceeding to production
- Document any issues encountered and their resolutions

:::important
The value of a staging environment cannot be overstated. It provides a safe space to identify and resolve issues before they impact production workloads.
:::

### Verify High Availability

If you require minimal downtime during upgrades:

- Ensure your applications are deployed with proper HA configurations
- Understand which UDS Core components may experience brief unavailability during upgrades
- Plan maintenance windows accordingly for components that cannot be upgraded without interruption

## Upgrade Process

### Update the UDS Core Bundle

Typically the actual version update is as easy as updating your version reference in a `uds-bundle.yaml`. For example if you had a bundle like the below:

```yaml
packages:
  - name: core
    repository: ghcr.io/defenseunicorns/packages/uds/core
    ref: 0.9.0-upstream
```

Upgrading to 0.10.0 would be as easy as updating this to:

```yaml
packages:
  - name: core
    repository: ghcr.io/defenseunicorns/packages/uds/core
    ref: 0.10.0-upstream
```

> Try to avoid other concurrent package upgrades (i.e. zarf init, other UDS Packages) or larger changes like switching between flavors unless you have restrictive maintenance windows. Where possible it is often better to perform these upgrades independently to identify and troubleshoot if issues occur. 

Reviewer 2: nit: maybe some clearer sentence flow on this one?
```suggestion
Try to avoid other concurrent package upgrades (e.g., zarf init or other UDS packages) or larger changes, such as switching between flavors, unless you have restrictive maintenance windows. Where possible, it is often better to perform these upgrades independently to simplify troubleshooting if issues occur.
```

### Update Configurations

Before creating the new bundle, update configuration as needed:

1. **UDS Core Configuration Changes**:
   - Review any changes required for UDS Core custom resources
   - Review any values changes to UDS Core helm charts and zarf variables

2. **Upstream Tool Configuration Changes**:

>    - Review and upstream tooling release notes especially if there are major version updates

Reviewer 2: ```suggestion
   - Review the release notes for upstream tools, especially if major version updates are included
```

   - Where necessary update bundle overrides based on any helm chart values changes

### Build and Deploy Bundle

Once you have updated the version reference for UDS Core and made any necessary configuration changes, create the updated bundle and deploy it. The example below is the bare minimum to create and deploy:

```console
# Create the bundle
uds create <path/to/uds-bundle.yaml> --confirm

# Deploy the bundle
uds deploy <path/to/bundle.tar.zst> --confirm
```

Depending on your configuration and process this may have additional steps with variables, dynamic environment configuration, etc.

## Post-Upgrade Verification

After the bundle deployment completes, verify the health and functionality of your environment:

1. **Verify UDS Core Components**:

>   - The UDS Core deployment will perform basic health checking for you

Reviewer 2: nit:
```suggestion
  - The UDS Core deployment performs basic health checks automatically
```

  - Additionally confirm all UDS Core components are accessible at their endpoints, with SSO login

2. **Verify Mission Applications**:
  - Check that your mission apps are still running and healthy
  - Validate endpoint accessibility and proper configuration (i.e. monitoring, SSO working as expected)

## Upgrade Troubleshooting and Rollbacks

UDS Core does not officially test or support rollback procedures. Individual opensource applications included in the UDS Core platform may not behave well during a rollback.

Rather than attempting a rollback we recommend the following approaches:

1. **Roll Forward**: Address issues by applying fixes or configuration changes to the current version.

> 1. **Manual Intervention**: Where necessary perform manual "one-time" upgrade fixes to restore access. If there are persistent issues, these should reported as [GitHub Issues](https://github.com/defenseunicorns/uds-core/issues) for the team to address going forward.

Reviewer 2: ```suggestion
1. **Manual Intervention**: Where necessary perform manual "one-time" upgrade fixes to restore access. If there are persistent issues, these should be reported as [GitHub Issues](https://github.com/defenseunicorns/uds-core/issues) for the team to address going forward.
```

1. **Restore from Backup**: In critical situations, consider restoring from backups rather than attempting a version rollback.

> With any upgrade issues it is also important to:

Reviewer 2: nit:
```suggestion
If you encounter upgrade issues, it's important to:
```
